libtenzir/builtins/operators/compress_decompress.cpp

```diff
@@ -0,0 +1,387 @@
+        continue;
+      }
+      in_buffer.consume(std::move(bytes));
+      while (in_buffer.size() > 0) {
+        auto result = decompressor.ValueUnsafe()->Decompress(
+          in_buffer.size(), in_buffer.data(),
+          detail::narrow_cast<int64_t>(out_buffer.size()), out_buffer.data());
+        if (not result.ok()) {
+          diagnostic::error("failed to decompress: {}",
+                            result.status().ToString())
+            .emit(ctrl.diagnostics());
+          co_return;
+        }
+        in_buffer.drop_front_n(result->bytes_read);
+        // Some decompressors signal that they need a larger output buffer to
+        // write into. We already start with a rather large one, but in case
+        // that isn't enough we just double the size until the decompressor
+        // stops complaining.
+        if (result->need_more_output) [[unlikely]] {
+          out_buffer.resize(out_buffer.size() * 2);
+        }
+        if (result->bytes_written > 0) {
+          TENZIR_ASSERT(result->bytes_written
+                        < detail::narrow_cast<int64_t>(out_buffer.size()));
+          co_yield chunk::copy(
+            as_bytes(out_buffer).subspan(0, result->bytes_written));
+        } else {
+          break;
+        }
+        // In case the input contains multiple concatenated compressed streams,
+        // we gracefully reset the decompressor.
+        if (decompressor.ValueUnsafe()->IsFinished()) [[unlikely]] {
+          const auto reset_status = decompressor.ValueUnsafe()->Reset();
+          if (not reset_status.ok()) {
+            diagnostic::error("failed to reset decompressor: {}",
+                              reset_status.ToString())
+              .emit(ctrl.diagnostics());
+            co_return;
+          }
+        }
+      }
+    }
+    if (not decompressor.ValueUnsafe()->IsFinished()) {
+      TENZIR_VERBOSE(
+        "decompressor is not finished, but end of input is reached");
+    }
+  }
+
+  auto name() const -> std::string override {
+    return "decompress";
+  }
+
+  auto optimize(expression const& filter, event_order order) const
+    -> optimize_result override {
+    (void)filter;
+    (void)order;
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, decompress_operator& x) -> bool {
+    return f.object(x)
+      .pretty_name("decompress_operator")
+      .fields(f.field("args", x.args_));
+  }
+
+private:
+  operator_args args_ = {};
+};
+
+class compress_plugin final
+  : public virtual operator_plugin<compress_operator> {
+public:
+  auto signature() const -> operator_signature override {
+    return {.transformation = true};
+  }
+
+  auto parse_operator(parser_interface& p) const -> operator_ptr override {
+    auto parser
+      = argument_parser{"decompress", "https://docs.tenzir.com/next/operators/"
+                                      "transformations/compress"};
+    auto args = operator_args{};
+    parser.add(args.type, "<type>");
```

Comment:
Maybe you can improve diagnostics output if you replaced `<type>` with a string specifying all valid values: ```[brotli | bz2 | gzip | lz4 | zstd]```

Maybe a next step for the diagnostics would be allowing a generic meta string `<type>`, with additional optional specification: `must be any of: brotli, ...` @jachris 

Also, it's `<codec>` in the docs.

Reply:
I have the list of supported codecs in the error message when trying to create the codec, which I think works well enough.